File: Python/100. Same Tree.py
'''
1.递归
2.迭代
'''


# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

# class Solution:
#     def isSameTree(self, p: Optional[TreeNode], q: Optional[TreeNode]) -> bool:
#         if not p and not q:
#             return True
#         if not p or not q or (p.val != q.val):
#             return False
#         left = self.isSameTree(p.left, q.left)
#         right = self.isSameTree(p.right, q.right)
#         return left and right

class Solution:
    def isSameTree(self, p: Optional[TreeNode], q: Optional[TreeNode]) -> bool:
        que = [p, q]
        while que:
            left = que.pop(0)
            right = que.pop(0)
            if (not left and not right):
                continue
            if (not left or not right or left.val != right.val):
                return False
            que.append(left.left)
            que.append(right.left)
            que.append(left.right)
            que.append(right.right)
        return True

File: C++/100. Same Tree.cpp
/*
1.µÝ¹é
2.µü´ú
*/
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

 // class Solution {
 // public:
 //     bool isSameTree(TreeNode* p, TreeNode* q) {
 //         if (!p && !q) return true;
 //         if (!p || !q || p->val != q->val) return false;
 //         bool left = isSameTree(p->left, q->left);
 //         bool right = isSameTree(p->right, q->right);
 //         return left&&right;
 //     }
 // };

class Solution {
public:
    bool isSameTree(TreeNode* p, TreeNode* q) {
        queue<TreeNode*> que;
        que.push(p);
        que.push(q);
        while (!que.empty()) {
            TreeNode* left = que.front(); que.pop();
            TreeNode* right = que.front(); que.pop();
            if (!left && !right) continue;
            if (!left || !right || left->val != right->val) return false;
            que.push(left->left);
            que.push(right->left);
            que.push(left->right);
            que.push(right->right);
        }
        return true;
    }
};